File: Freestylecoding.Math.FSharp.Tests/Natural/ComparisonOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public Equality() =
    [<Theory>]
    [<InlineData( 0u, 0u, true )>]
    [<InlineData( 0u, 1u, false )>]
    [<InlineData( 1u, 0u, false )>]
    [<InlineData( 1u, 1u, true )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_Equality( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.False( Natural.op_Equality( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.False( Natural.op_Equality( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )

type public GreaterThan() =
    [<Theory>]
    [<InlineData( 0u, 1u, false )>]
    [<InlineData( 1u, 0u, true )>]
    [<InlineData( 1u, 1u, false )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_GreaterThan( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.True( Natural.op_GreaterThan( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.False( Natural.op_GreaterThan( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )
    
    [<Fact>]
    member public this.CascadeGreaterThan () =
        Assert.True( Natural.op_GreaterThan( Natural( [1u; 1u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeEqual () =
        Assert.False( Natural.op_GreaterThan( Natural( [1u; 0u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeLessThan () =
        Assert.False( Natural.op_GreaterThan( Natural( [1u; 0u] ), Natural( [1u; 1u] ) ) )

type public LessThan() =
    [<Theory>]
    [<InlineData( 0u, 1u, true )>]
    [<InlineData( 1u, 0u, false )>]
    [<InlineData( 1u, 1u, false )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_LessThan( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.False( Natural.op_LessThan( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.True( Natural.op_LessThan( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )
    
    [<Fact>]
    member public this.CascadeGreaterThan () =
        Assert.False( Natural.op_LessThan( Natural( [1u; 1u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeEqual () =
        Assert.False( Natural.op_LessThan( Natural( [1u; 0u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeLessThan () =
        Assert.True( Natural.op_LessThan( Natural( [1u; 0u] ), Natural( [1u; 1u] ) ) )

type public GreaterThanOrEqual() =
    [<Theory>]
    [<InlineData( 0u, 1u, false )>]
    [<InlineData( 1u, 0u, true )>]
    [<InlineData( 1u, 1u, true )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_GreaterThanOrEqual( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.True( Natural.op_GreaterThanOrEqual( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.False( Natural.op_GreaterThanOrEqual( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )
    
    [<Fact>]
    member public this.CascadeGreaterThan () =
        Assert.True( Natural.op_GreaterThanOrEqual( Natural( [1u; 1u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeEqual () =
        Assert.True( Natural.op_GreaterThanOrEqual( Natural( [1u; 0u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeLessThan () =
        Assert.False( Natural.op_GreaterThanOrEqual( Natural( [1u; 0u] ), Natural( [1u; 1u] ) ) )

type public LessThanOrEqual() =
    [<Theory>]
    [<InlineData( 0u, 1u, true )>]
    [<InlineData( 1u, 0u, false )>]
    [<InlineData( 1u, 1u, true )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_LessThanOrEqual( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.False( Natural.op_LessThanOrEqual( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.True( Natural.op_LessThanOrEqual( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )
    
    [<Fact>]
    member public this.CascadeGreaterThan () =
        Assert.False( Natural.op_LessThanOrEqual( Natural( [1u; 1u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeEqual () =
        Assert.True( Natural.op_LessThanOrEqual( Natural( [1u; 0u] ), Natural( [1u; 0u] ) ) )
    
    [<Fact>]
    member public this.CascadeLessThan () =
        Assert.True( Natural.op_LessThanOrEqual( Natural( [1u; 0u] ), Natural( [1u; 1u] ) ) )

type public Inequality() =
    [<Theory>]
    [<InlineData( 0u, 0u, false )>]
    [<InlineData( 0u, 1u, true )>]
    [<InlineData( 1u, 0u, true )>]
    [<InlineData( 1u, 1u, false )>]
    member public this.Sanity left right expected =
        Assert.Equal( expected, Natural.op_Inequality( Natural( [left] ), Natural( [right] ) ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.True( Natural.op_Inequality( Natural( [0xBADu; 0xDEADBEEFu] ), Natural( [0xDEADBEEFu] ) ) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.True( Natural.op_Inequality( Natural( [0xDEADBEEFu] ), Natural( [0xBADu; 0xDEADBEEFu] ) ) )

File: Freestylecoding.Math.CSharp.Tests/Natural/ComparisonOperators.cs
using Xunit;

namespace Freestylecoding.Math.CSharp.Tests.Natural;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class Equality {
	[Theory]
	[InlineData( 0u, 0u, true )]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 1u, 1u, true )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			new Natural( left ) == new Natural( right )
		);

	[Fact]
	public void BiggerLeft() =>
		Assert.False( FListNatural( 0xBADu, 0xDEADBEEFu ) == new Natural( 0xDEADBEEFu ) );

	[Fact]
	public void BiggerRight() =>
		Assert.False( new Natural( 0xDEADBEEFu ) == FListNatural( 0xBADu, 0xDEADBEEFu ) );
}

public class GreaterThan {
	[Theory]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 1u, 1u, false )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal( expected, new Natural( left ) > new Natural( right ) );

	[Fact]
	public void BiggerLeft() =>
		Assert.True( FListNatural( 0xBADu, 0xDEADBEEFu ) > new Natural( 0xDEADBEEFu ) );

	[Fact]
	public void BiggerRight() =>
		Assert.False( new Natural( 0xDEADBEEFu ) > FListNatural( 0xBADu, 0xDEADBEEFu ) );

	[Fact]
	public void CascadeGreaterThan() =>
		Assert.True( FListNatural( 1u, 1u ) > FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeEqual() =>
		Assert.False( FListNatural( 1u, 0u ) > FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeLessThan() =>
		Assert.False( FListNatural( 1u, 0u ) > FListNatural( 1u, 1u ) );
}

public class LessThan {
	[Theory]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 1u, 1u, false )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal( expected, new Natural( left ) < new Natural( right ) );

	[Fact]
	public void BiggerLeft() =>
		Assert.False( FListNatural( 0xBADu, 0xDEADBEEFu ) < new Natural( 0xDEADBEEFu ) );

	[Fact]
	public void BiggerRight() =>
		Assert.True( new Natural( 0xDEADBEEFu ) < FListNatural( 0xBADu, 0xDEADBEEFu ) );

	[Fact]
	public void CascadeGreaterThan() =>
		Assert.False( FListNatural( 1u, 1u ) < FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeEqual() =>
		Assert.False( FListNatural( 1u, 0u ) < FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeLessThan() =>
		Assert.True( FListNatural( 1u, 0u ) < FListNatural( 1u, 1u ) );
}

public class GreaterThanOrEqual {
	[Theory]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 1u, 1u, true )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal( expected, new Natural( left ) >= new Natural( right ) );

	[Fact]
	public void BiggerLeft() =>
		Assert.True( FListNatural( 0xBADu, 0xDEADBEEFu ) >= new Natural( 0xDEADBEEFu ) );

	[Fact]
	public void BiggerRight() =>
		Assert.False( new Natural( 0xDEADBEEFu ) >= FListNatural( 0xBADu, 0xDEADBEEFu ) );

	[Fact]
	public void CascadeGreaterThan() =>
		Assert.True( FListNatural( 1u, 1u ) >= FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeEqual() =>
		Assert.True( FListNatural( 1u, 0u ) >= FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeLessThan() =>
		Assert.False( FListNatural( 1u, 0u ) >= FListNatural( 1u, 1u ) );
}

public class LessThanOrEqual {
	[Theory]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 1u, 1u, true )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal( expected, new Natural( left ) <= new Natural( right ) );

	[Fact]
	public void BiggerLeft() =>
		Assert.False( FListNatural( 0xBADu, 0xDEADBEEFu ) <= new Natural( 0xDEADBEEFu ) );

	[Fact]
	public void BiggerRight() =>
		Assert.True( new Natural( 0xDEADBEEFu ) <= FListNatural( 0xBADu, 0xDEADBEEFu ) );

	[Fact]
	public void CascadeGreaterThan() =>
		Assert.False( FListNatural( 1u, 1u ) <= FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeEqual() =>
		Assert.True( FListNatural( 1u, 0u ) <= FListNatural( 1u, 0u ) );

	[Fact]
	public void CascadeLessThan() =>
		Assert.True( FListNatural( 1u, 0u ) <= FListNatural( 1u, 1u ) );
}

public class Inequality {
	[Theory]
	[InlineData( 0u, 0u, false )]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 1u, 1u, false )]
	public void Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal( expected, new Natural( left ) != new Natural( right ) );

	[Fact]
	public void BiggerLeft() =>
		Assert.True( FListNatural( 0xBADu, 0xDEADBEEFu ) != new Natural( 0xDEADBEEFu ) );

    [Fact]
	public void BiggerRight() =>
		Assert.True( new Natural( 0xDEADBEEFu ) != FListNatural( 0xBADu, 0xDEADBEEFu ) );
}